### ISSUE TITLE
IHasMeta does not work

### ISSUE THREAD

**Issue by bart-degreed** [json-api-dotnet/JsonApiDotNetCore#737]
Example:

```csharp
    public class Person : Identifiable, IHasMeta
    {
        [Attr]
        public string FirstName { get; set; }
        
        [Attr]
        public string LastName { get; set; }

        public Dictionary<string, object> GetMeta()
        {
            return new Dictionary<string, object>
            {
                {"copyright", "Copyright 2018 Example Corp."},
                {"authors", new[] {"John Doe"}}
            };
        }
    }
```

```
curl https://localhost:44354/people
```

```json
{"links":{"self":"https://localhost:44354///people"},"data":[{"type":"people","id":"1","attributes":{"firstName":"John","lastName":"Simmons"},"links":{"self":"https://localhost:44354///people/1"}},{"type":"people","id":"2","attributes":{"firstName":"Pete","lastName":"Bell"},"links":{"self":"https://localhost:44354///people/2"}}]}
```

**Comment by bart-degreed** [comment]
Same for `IRequestMeta`:

```csharp
public class RequestMetaService : IRequestMeta
{
    public Dictionary<string, object> GetMeta()
    {
        return new Dictionary<string, object>
        {
            {"copyright", "Copyright 2018 Example Corp."},
            {"authors", new string[] {"John Doe"}}
        };
    }
}
```

**Comment by bart-degreed** [comment]
Current behavior is that `IRequestMeta` adds top-level meta to the response document unconditionally, while `IHasMeta` (which was replaced with `IResourceDefinition.GetMeta` in #832) adds top-level meta only when the requested resource type matches.

We've discussed to change this, because:
- The json:api spec allows meta to occur per resource _instance_ (non top-level), which we currently do not support but seems more useful.
- We suspect that the matching of resource type does not work properly. For example, there is no obvious requested resource type for a request like `/blogs/1/relationships/articles`, because it functionally targets a relationship between two resources of potentially different types.

Our goal is to change this to:
1. Support top-level meta unconditionally (typically used for API version, copyright notice etc) -- as-is
2. Support nested meta per resource instance (use case: some articles may have a discount, product recalls etc) -- requires that the resource instance is passed to `IResourceDefinition.GetMeta` calls and it is invoked (multiple times) during the serialization process.
